src/jit/alloc.cpp

```diff
@@ -1,411 +1,389 @@
-
-void        norls_allocator::nraFree(void)
+void ArenaAllocator::destroy()
 {
-    /* Free all of the allocated pages */
-
-    while   (nraPageList)
+    // Free all of the allocated pages
+    for (PageDescriptor* page = m_firstPage, *next; page != nullptr; page = next)
     {
-        norls_pagdesc * temp;
+        next = page->m_next;
+        freeHostMemory(page);
+    }
 
-        temp = nraPageList;
-               nraPageList = temp->nrpNextPage;
+    // Clear out the allocator's fields
+    m_firstPage = nullptr;
+    m_lastPage = nullptr;
+    m_nextFreeByte = nullptr;
+    m_lastFreeByte = nullptr;
+}
 
-        nraVirtualFree(temp, 0, MEM_RELEASE);
-    }
+void ArenaAllocator::mark(MarkDescriptor& mark)
+{
+    mark.m_page = m_lastPage;
+    mark.m_next = m_nextFreeByte;
+    mark.m_last = m_lastFreeByte;
 }
 
-// This method walks the nraPageList backward and release the pages.
+// This method may walk the m_firstPage backward and release the pages.
 // Be careful no other thread is doing nraFree as the same time.
 // Otherwise, the page specified by temp could be double-freed (VSW 600919).
-void        norls_allocator::nraToss(nraMarkDsc &mark)
+void ArenaAllocator::reset(MarkDescriptor& mark)
 {
-    void    *   last = mark.nmPage;
+    // If the active page hasn't changed, just reset the position into the
+    // page and return.
+    if (m_lastPage == mark.m_page)
+    {
+        m_nextFreeByte = mark.m_next;
+        m_lastFreeByte = mark.m_last;
+        return;
+    }
+
+    // Otherwise, free any new pages that were added.
+    void* last = mark.m_page;
 
-    if  (!last)
+    if (last == nullptr)
     {
-        if  (!nraPageList)
+        if (m_firstPage == nullptr)
+        {
             return;
+        }
 
-        nraFreeNext  = nraPageList->nrpContents;
-        nraFreeLast  = nraPageList->nrpPageSize + (BYTE *)nraPageList;
-
+        m_nextFreeByte = m_firstPage->m_contents;
+        m_lastFreeByte = m_firstPage->m_pageBytes + (BYTE*)m_firstPage;
         return;
     }
 
-    /* Free up all the new pages we've added at the end of the list */
-
-    while (nraPageLast != last)
+    while (m_lastPage != last)
     {
-        norls_pagdesc * temp;
-
-        /* Remove the last page from the end of the list */
+        // Remove the last page from the end of the list
+        PageDescriptor* temp = m_lastPage;
+        m_lastPage = temp->m_previous;
 
-        temp = nraPageLast;
-               nraPageLast = temp->nrpPrevPage;
+        // The new last page has no next page
+        m_lastPage->m_next = nullptr;
 
-        /* The new last page has no 'next' page */
+        freeHostMemory(temp);
+    }
 
-        nraPageLast->nrpNextPage = 0;
+    m_nextFreeByte = mark.m_next;
+    m_lastFreeByte = mark.m_last;
+}
 
-        nraVirtualFree(temp, 0, MEM_RELEASE);
+void* ArenaAllocator::allocateHostMemory(size_t size)
+{
+#if defined(DEBUG)
+    if (!useHostAllocator())
+    {
+#undef GetProcessHeap
+#undef HeapAlloc
```

Comment:
The CLR typically defines symbols that prevent accidental uses of the OS allocator, since allocation requests are typically required to use the hosting API. I've added a comment to that effect and consolidated the `#undef`s.